core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -158,11 +160,66 @@ RelSubset getOrCreateSubset(
       final VolcanoPlanner planner =
           (VolcanoPlanner) cluster.getPlanner();
 
+      // Converters from newly introduced subset to all the remaining one, only if possible to
+      // convert.  No point adding converters if it is not possible.
+      for (RelSubset other : subsets) {
+        if (other.getConvention() == Convention.NONE) {
+          continue;
+        }
+
+        assert other.getTraitSet().size() == subset.getTraitSet().size();
+
+        final ImmutableList<RelTrait> otherTrait =
+                other.getTraitSet().difference(subset.getTraitSet());
+        final ImmutableList<RelTrait> thisTrait =
+                subset.getTraitSet().difference(other.getTraitSet());
+        assert otherTrait.size() == thisTrait.size();
+
+        if (otherTrait.size() == 1 && thisTrait.size() == 1) {
+          RelTraitDef traitDef = otherTrait.get(0).getTraitDef();
+          boolean canConvert = traitDef.canConvert(
+                  cluster.getPlanner(), thisTrait.get(0), otherTrait.get(0));
+
+          if (canConvert) {
+            final AbstractConverter converter =
+                    new AbstractConverter(cluster, subset, traitDef, other.getTraitSet());
+            planner.register(converter, other);
+          }
+        }
+      }
+
       subsets.add(subset);
 
-      if (planner.root != null
-          && planner.root.set == this) {
-        planner.ensureRootConverters();
+      // Converters from newly introduced subset to all the remaining one, only if possible to
+      // convert.  No point adding converters if it is not possible.
+      for (RelSubset other : subsets) {
+        if (other == subset) {
+          continue;
+        }
+
+        if (other.getConvention() == Convention.NONE) {
```

Comment:
I think there is another bug, which I need to dig a bit more to figure out, but when I remove this check (for Convention.NONE), we fail with IndexOutOfBounds exception in RelCollationTraitDef.convert().  For example, in one of the JdbcTest (testModelWithInvalidModifiableView), we fail to convert from NONE.[1,0,2] to NONE.[0,2,3].  I will look into this and file another bug instead of trying to fix it here.

java.lang.AssertionError: Internal error: Error while applying rule ExpandConversionRule, args [rel#192:AbstractConverter.NONE.[0, 2, 3](input=rel#191:Subset#4.NONE.[1, 0, 2],convention=NONE,sort=[0, 2, 3])]

	at org.apache.calcite.util.Util.newInternal(Util.java:792)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:237)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.findBestExp(VolcanoPlanner.java:825)
	at org.apache.calcite.tools.Programs$5.run(Programs.java:300)
	at org.apache.calcite.tools.Programs$SequenceProgram.run(Programs.java:351)
	at org.apache.calcite.prepare.Prepare.optimize(Prepare.java:153)
	at org.apache.calcite.prepare.Prepare.prepareSql(Prepare.java:276)
	at org.apache.calcite.prepare.Prepare.prepareSql(Prepare.java:193)
	at org.apache.calcite.prepare.CalcitePrepareImpl.prepare2_(CalcitePrepareImpl.java:720)
	at org.apache.calcite.prepare.CalcitePrepareImpl.prepare_(CalcitePrepareImpl.java:587)
	at org.apache.calcite.prepare.CalcitePrepareImpl.prepareSql(CalcitePrepareImpl.java:557)
	at org.apache.calcite.jdbc.CalciteConnectionImpl.parseQuery(CalciteConnectionImpl.java:214)
	at org.apache.calcite.jdbc.CalciteMetaImpl.prepareAndExecute(CalciteMetaImpl.java:573)
	at org.apache.calcite.avatica.AvaticaConnection.prepareAndExecuteInternal(AvaticaConnection.java:581)
	at org.apache.calcite.avatica.AvaticaStatement.executeInternal(AvaticaStatement.java:135)
	at org.apache.calcite.avatica.AvaticaStatement.executeQuery(AvaticaStatement.java:186)
	at org.apache.calcite.test.CalciteAssert.assertQuery(CalciteAssert.java:472)
	at org.apache.calcite.test.CalciteAssert$AssertQuery.returns(CalciteAssert.java:1211)
	at org.apache.calcite.test.CalciteAssert$AssertQuery.explainMatches(CalciteAssert.java:1293)
	at org.apache.calcite.test.CalciteAssert$AssertQuery.explainContains(CalciteAssert.java:1288)
	at org.apache.calcite.test.JdbcTest.testModelWithInvalidModifiableView(JdbcTest.java:355)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.lang.IndexOutOfBoundsException: index (3) must be less than size (3)
	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:310)
	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:292)
	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:65)
	at org.apache.calcite.rex.RexBuilder.makeInputRef(RexBuilder.java:797)
	at org.apache.calcite.rel.core.Sort.<init>(Sort.java:103)
	at org.apache.calcite.rel.logical.LogicalSort.<init>(LogicalSort.java:37)
	at org.apache.calcite.rel.logical.LogicalSort.create(LogicalSort.java:63)
	at org.apache.calcite.rel.RelCollationTraitDef.convert(RelCollationTraitDef.java:75)
	at org.apache.calcite.rel.RelCollationTraitDef.convert(RelCollationTraitDef.java:39)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.changeTraitsUsingConverters(VolcanoPlanner.java:1218)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.changeTraitsUsingConverters(VolcanoPlanner.java:1310)
	at org.apache.calcite.plan.volcano.AbstractConverter$ExpandConversionRule.onMatch(AbstractConverter.java:116)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:214)
	... 45 more


Process finished with exit code 255